src/types.rs

```diff
@@ -370,9 +399,333 @@ impl From<ReturnCode> for u8 {
+    /// Instance ID
+    pub instance_id: InstanceId,
+    /// Major version
+    pub major_version: MajorVersion,
+    /// Minor version
+    pub minor_version: MinorVersion,
+    /// TTL
+    pub ttl: u24,
+    /// Options
+    pub options: SdOptionRef,
+}
+
+impl SdServiceEntry {
+    /// Has TTL
+    pub fn has_ttl(&self) -> bool {
+        self.ttl != u24::new(0)
+    }
+}
+
+/// SD event group entry
+#[derive(Debug, PartialEq)]
+pub struct SdEventgroupEntry {
+    /// Service ID
+    pub service_id: ServiceId,
+    /// Instance ID
+    pub instance_id: InstanceId,
+    /// Eventgroup ID
+    pub eventgroup_id: EventgroupId,
+    /// Major version
+    pub major_version: MajorVersion,
+    /// TTL
+    pub ttl: u24,
+    /// Options
+    pub options: SdOptionRef,
+}
+
+impl SdEventgroupEntry {
+    /// Has TTL
+    pub fn has_ttl(&self) -> bool {
+        self.ttl != u24::new(0)
+    }
+}
+
+/// Instance ID
+pub type InstanceId = u16;
+
+/// Eventgroup ID
+pub type EventgroupId = u16;
+
+/// Major version
+pub type MajorVersion = u8;
+
+/// Minor version
+pub type MinorVersion = u32;
+
+/// SD Option reference
+#[allow(missing_docs)]
+#[derive(Debug, PartialEq)]
+pub struct SdOptionRef {
+    pub index1: u8,
+    pub index2: u8,
+    pub num1: u8,
+    pub num2: u8,
+}
+
+/// SD option
+#[derive(Debug, PartialEq)]
+#[allow(missing_docs)]
+pub enum SdOption {
+    Ip4Unicast(SdEndpointOption),
+    Ip4Multicast(SdEndpointOption),
+    Ip6Unicast(SdEndpointOption),
+    Ip6Multicast(SdEndpointOption),
+}
+
+impl SdOption {
+    /// Returns true if this is a IP option
+    pub fn is_ip_option(option_type: u8) -> bool {
+        Self::is_ip4_option(option_type) || Self::is_ip6_option(option_type)
+    }
+
+    /// Returns true if this is a IPv4 option
+    pub fn is_ip4_option(option_type: u8) -> bool {
+        matches!(option_type, 0x04 | 0x14)
+    }
+
+    /// Returns true if this is a IPv6 option
+    pub fn is_ip6_option(option_type: u8) -> bool {
+        matches!(option_type, 0x06 | 0x16)
+    }
+
+    /// Returns true if this is a multicast option
+    pub fn is_multicast_option(option_type: u8) -> bool {
+        matches!(option_type, 0x14 | 0x16)
+    }
+
+    /// Construct a new SD option from type and endpoint
+    pub fn from(option_type: u8, option: SdEndpointOption) -> Result<Self, Error> {
+        use SdOption::*;
+        match option_type {
+            0x04 => Ok(Ip4Unicast(option)),
+            0x14 => Ok(Ip4Multicast(option)),
+            0x06 => Ok(Ip6Unicast(option)),
+            0x16 => Ok(Ip6Multicast(option)),
+            option_type => Err(Error::UnknownSdOption(option_type)),
+        }
+    }
+
+    /// Type
+    pub fn r#type(&self) -> u8 {
+        use SdOption::*;
+        match self {
+            Ip4Unicast(_) => 0x04,
+            Ip4Multicast(_) => 0x14,
+            Ip6Unicast(_) => 0x06,
+            Ip6Multicast(_) => 0x16,
+        }
+    }
+
+    /// Length
+    pub fn len(&self) -> usize {
```

Comment:
enum `SdOption` has a public `len` method, but no `is_empty` method

Reply:
renamed to length()